### PR TITLE
use tokenId to get to cosoul art page

### PR DESCRIPTION
**Pull request by sshmm**
### <samp>🤖 Generated by Copilot at d204a93</samp>

This pull request adds a new feature to the coordinape app that allows users to view and share the cosoul art they created or received as NFTs. It implements a new API endpoint for fetching the cosoul art data by token id and a new component that renders the cosoul art with animation. It also refactors and improves the existing code for the cosoul feature, adding validation and error handling.

### <samp>🤖 Generated by Copilot at d204a93</samp>

> _`CoSoulArt` shines_
> _New API endpoint added_
> _Winter of refactor_

## Why

Update CoSoul Art Only View page to use DB instead of query params
Change the animation_url to on the metadata api to `${WEB_APP_BASE_URL}/cosoul/art/${coSoulData.token_id};`
And then update the cosoul art page to fetch the data from the token_id and render properly.

## Screenshots (if appropriate):
image

### <samp>🤖 Generated by Copilot at d204a93</samp>

*  Validate the address parameter in the `api/cosoul/[address].ts` file using the `isAddress` function and throw a `NotFoundError` if invalid or not found ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-c05b4221a8fa5c2d52523c14cdf221c5dc469b8ee2f4a6dd47a3cc32babf6852L4-R8), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-c05b4221a8fa5c2d52523c14cdf221c5dc469b8ee2f4a6dd47a3cc32babf6852L22-R26), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-c05b4221a8fa5c2d52523c14cdf221c5dc469b8ee2f4a6dd47a3cc32babf6852L57))
* Add a new API endpoint for fetching the cosoul art data based on the token id in the `api/cosoul/art/[artTokenId].ts` file and export the `CosoulArtData` type ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-8a513163f8449f612956f9c609b9b079493b1c71f86cd13dc2f8d3902724387aR1-R66))
* Fetch the token id of the cosoul from the database and use it to construct the `animation_url` field in the `api/cosoul/metadata/[tokenId].ts` file and throw a `NotFoundError` if not found ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-e8f0277d6e465db91324fc6ac3dce8949d80850e95e0bb69a92a326a9b0799c8L18-R20), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-e8f0277d6e465db91324fc6ac3dce8949d80850e95e0bb69a92a326a9b0799c8R43), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-e8f0277d6e465db91324fc6ac3dce8949d80850e95e0bb69a92a326a9b0799c8R58-R62), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-e8f0277d6e465db91324fc6ac3dce8949d80850e95e0bb69a92a326a9b0799c8L67-R70))
* Register the new API endpoint for fetching the cosoul art data based on the token id in the local development server in the `scripts/serve_dev.ts` file ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-dd65c0157396f6335fc1f86164d9ffeb75c19d3c2e7fb5ebcd684e9f1a41f5b7R7), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-dd65c0157396f6335fc1f86164d9ffeb75c19d3c2e7fb5ebcd684e9f1a41f5b7L77-R84))
* Remove the URL params for testing and use the props passed from the parent component in the `CoSoulArt` component in the `src/features/cosoul/art/CoSoulArt.tsx` file and remove the `console.log` statement ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-0d32f46ad080049e96febbbed420394d392db3af9f64f964210bee3bab9d91feL8-R12), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-0d32f46ad080049e96febbbed420394d392db3af9f64f964210bee3bab9d91feL40-R32))
* Add a new component `CoSoulArtPublic` that renders the `CoSoulArt` component with the data fetched from the new API endpoint for fetching the cosoul art data based on the token id in the `src/features/cosoul/art/CoSoulArtPublic.tsx` file ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-92b5ca8f39f680ccbd0ad803deb384daa76a397984fbb9bd6226d9552f5d08ffR1-R36))
* Change the `cosoulArt` field in the `paths` object to a function that takes a token id as a parameter and returns the URL path for the new API endpoint for fetching the cosoul art data based on the token id in the `src/routes/paths.ts` file ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-26212c7e58b310fe67e20a7cd969dce535cc31c1f843f741c71641dddc7320ddL73-R73))
* Change the `Route` element for the `cosoulArt` path to use the `CoSoulArtPublic` component wrapped in a `RequireAuth` component and the `paths.cosoulArt` function with the `:tokenId` parameter in the `src/routes/routes.tsx` file ([link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-0005a5feea8701ddb5aef0d57e1a5a03c36dae6cd62fc3ecf35fd15a88bb5c03L9-R9), [link](https://github.com/coordinape/coordinape/pull/2210/files?diff=unified&w=0#diff-0005a5feea8701ddb5aef0d57e1a5a03c36dae6cd62fc3ecf35fd15a88bb5c03L150-R157))

### CODE REVIEW

File: api/cosoul/art/[artTokenId].ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+const CACHE_SECONDS = 60 * 5;
+export default async function handler(req: VercelRequest, res: VercelResponse) {
+  try {
+    let artTokenId: number | undefined;
+    if (typeof req.query.artTokenId == 'string') {
+      artTokenId = parseInt(req.query.artTokenId);
+    }
+
+    assert(artTokenId, 'no token Id provided');
```
Comment (crabsinger) [inline api/cosoul/art/[artTokenId].ts:17]: the assert will throw a 500, so this should be removed with the throw below so we return a 404 instead.

File: api/cosoul/art/[artTokenId].ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        {
+          pgive: true,
+          profile: {
+            address: true,
+          },
+        },
+      ],
+    },
+    {
+      operationName: 'cosoulApi__fetchMetadata',
```
Comment (crabsinger) [inline api/cosoul/art/[artTokenId].ts:50]: rename this i think,..

**Review by crabsinger — COMMENTED** [review]
a few changes

File: src/features/cosoul/art/CoSoulArtPublic.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+  const profileAddress = data?.address;
+  const userPgive = data?.pGive;
+
+  return (
+    <CoSoulArt
+      pGive={userPgive}
+      address={profileAddress}
+      animate={true}
+      showGui={true}
```
[the remark this answers is not in the document]
Comment (sshmm) [reply src/features/cosoul/art/CoSoulArtPublic.tsx:38]: animate and showGUI?

**Review by crabsinger — APPROVED** [review]
added a few fixes @sshmm - lgtm and merging

**Pull request merged**